### PR TITLE
[BEAM-14064] fix es io windowing

### PR DESCRIPTION
**Pull request by egalpin**
Ensure ElasticsearchIO#write output maintains input windows.

------------------------

Supersedes #17097

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [x] Update `CHANGES.md` with noteworthy changes.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by egalpin** [comment]
R: @je-ik @echauchot 

This PR is an alternative implementation from https://github.com/apache/beam/pull/17097 which maintains bundle-based batching while also ensuring windows are maintained across ElatichsearchIO#write.

File: sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java
Comment (je-ik) [inline sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:2324]: This looks like we still have the same issue with "outputWithTimestamp being before the current element timestamp", no? This can be fixed using the deprecated `getAllowedTimestampSkew`, but essentially it demonstrates the same deficiency.

Outputting with timestamp that might be (at least in theory) arbitrarily delayed after output watermark means that if user does a stateful (e.g. GBK) transform, some elements might get randomly dropped (although have been actually written).

I think we could work-around modifying the ElasticsearchIO transform, so that if the statful batching is not used, we return PDone instead. That way user cannot rely on any guarantees of what has been written. Or would have to ensure nothing gets dropped (setting allowedLateness to some high value, but that looks like highly suspicious side-effect of the transform).

I don't know what is correct solution here, it seems to me, that the root cause is the logical gap in stateless DoFn and how we work with watermarks there.

Comment (egalpin) [reply sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:2324]: You might be completely correct, I definitely don't feel I have a solid grasp just yet, but I'll give my interpretation of the root cause of the error that prompted this change and my understanding of how this PR might solve it.  I'm also operating under the assumption that the watermark is not advanced between elements of a single bundle, but instead advanced between bundles.

My understanding of the root cause is that it stems from the situation where there are elements in a bundle which belong to different windows.  This can even happen when using GIB because a bundle could have 2 elements where each element is a batch output from GIB.  Entities from each element in the bundle are then buffered.  If there were fewer entities in the bundle element than the max allowable number that can be buffered, the `@ProcessElement` completes with no output.  Then, another element from the bundle begins processing in `@ProcessElement`, this time with an element belonging to a window "ahead" in time than the window of the prior element.  Now any buffered entities cannot be output using `outputWithTimestamp` because the buffered entities have timestamps which are invalid for the window to which the element currently being processed by `@ProcessElement` belongs i.e. in the original Errors, the timestamps are not invalid for the pipeline with respect to the watermark, but are invalid when a different data from one window is output using another arbitrary window.

(this is my understanding, not necessarily fact/reality).

This PR might fix that issue by removing the error of outputting with invalid timestamps.  _If_ the watermark is held until an entire bundle is processed, then the existing error is not arising because the timestamps are invalid compared to the watermark, but strictly due to the window context from which they are attempting to be output.  By rewindowing into the GlobalWindow before all other processing, the Window context should always be the GlobalWindow for all elements in a bundle.  Then when we output with timestamp, the timestamps should no longer fail validation because they would be within the allowable bounds of [watermark, GlobalWindow.MAX_TIMESTAMP).  Then we re-window back into the original/input windowing strategy before putting dealing with everything in the global window.

Thoughts? Please correct above assumptions where I have things wrong.  CC @lukecwik as they deserve credit for the idea of reifying timestamps and global windowing -> rewindowing as a potential solution.

Comment (egalpin) [reply sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:2324]: Are there testing utilities that can synthesize a bundle to test this?

Comment (je-ik) [reply sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:2328]: Looking at the code in `SimpleDoFnRunner` I think that:
 a) `DoFnProcessContext.outputWithTimestamp` results in call to `checkTimestamp` (https://github.com/apache/beam/blob/096aeeff833679af3068eea2f00b280a589256e8/runners/core-java/src/main/java/org/apache/beam/runners/core/SimpleDoFnRunner.java#L409)
 b) the check will throw exception if the output timestamp is before the current element's timestamp (https://github.com/apache/beam/blob/096aeeff833679af3068eea2f00b280a589256e8/runners/core-java/src/main/java/org/apache/beam/runners/core/SimpleDoFnRunner.java#L444)
 
I *think* you can fix that by overriding `getAllowedTimestampSkew()` (returning some large value), and *provided* that - as you write - output watermark is held between bundles, then this should be correct. But I actually really doubt it is the case, I don't know how flink runner for instance would ensure that, bundle elements and watermarks are completely interleaved in this particular runner, because the runner does not have a "natural" concept of a bundle.

@robertwb maybe you could give some more insights?

> Are there testing utilities that can synthesize a bundle to test this?

You could probably use `TestStream` to verify the behavior on DirectRunner. I _believe_, that all elements that out add using `addElements` will end up in the same bundle.

Comment (egalpin) [reply sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:2328]: @je-ik thanks for the follow up, I really appreciate your time and input as I want to make the best fix possible here.  Agreed that I would need to add something to ResultFilterFn like:

```java
      @Override
      public Duration getAllowedTimestampSkew() {
        return Duration.millis(Long.MAX_VALUE);
      }
```

But I still wonder, as you point out, whether or not watermarks are updated during the processing of, or only between the processing of, bundles.  In the case of a runner like Flink with no notion of bundles, or in an imaginary runner where bundles were always of size 1, I suppose maybe the logic still holds true? Only after every bundle, i.e. after every element, has been processed can the watermark be updated.  ElasticsearchIO would be ok in that case as well, because `@FinishBundle` would output all elements before processing the element was done which would also imply being output before the watermark was updated?  Thinking out loud here...

Comment (je-ik) [reply sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:2333]: Yeah, that is about my doubts as well. Flink does not enforce size of bundle as 1, but instead handles them (optionally) in alignment with checkpoint barriers (and maxBundleSize). Watermarks flow with the data and (for stateless) DoFn I really don't see a way that it could guarantee holding the watermark between bundles.

Comment (egalpin) [reply sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:2333]: @je-ik What I meant to convey was that I think the solution as laid out in this PR may be resilient to runners where the watermark is updated after every element _iff_ we assume that this implies calling `@StartBundle` and `@FinishBundle` for each element (i.e. as if bundle size was enforced to be always 1).  Such an environment would cause all buffered elements to be flushed (via `@FinishBundle`) in the context of the window from which the elements originated before the watermark was updated.

It sounds like this notion is supported based on comments in the related dev@ thread[1].

[1] https://lists.apache.org/thread/7foy455spg43xo77zhrs62gc1m383t50

Comment (je-ik) [reply sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:2333]: I think we are going in cycles here. :) Maybe we could start a new dev@ thread to help clarifying things out. FlinkRunner definitely does not have bundles of size 1 (that is what would mean to call @StartBundle and @FinishBundle) before/after each element. That would make all "in-memory" batching void on such runner.
I think it implies one of two possibilities:
 a) either the watermark update only on boundaries of bundles is *not* part of the model, or
 b) FlinkRunner does not adhere to this model

Comment (je-ik) [reply sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:2333]: I created a thread for this: https://lists.apache.org/thread/10db7l9bhnhmo484myps723sfxtjwwmv

**Comment by je-ik** [comment]
Based on the discussion https://lists.apache.org/thread/10db7l9bhnhmo484myps723sfxtjwwmv I think that if we add the `getAllowedTimestampSkew()` call we can safely merge this. There will be some fix needed in FlinkRunner (and a \@ValidatesRunner test). I created https://issues.apache.org/jira/browse/BEAM-14196 to track this and will look into that.

**Comment by lukecwik** [comment]
> Based on the discussion https://lists.apache.org/thread/10db7l9bhnhmo484myps723sfxtjwwmv I think that if we add the `getAllowedTimestampSkew()` call we can safely merge this. There will be some fix needed in FlinkRunner (and a @ValidatesRunner test). I created https://issues.apache.org/jira/browse/BEAM-14196 to track this and will look into that.

It might make sense to fix flink and remove the `DoFnOperator.java` work around for Flink before we merge this.

**Comment by egalpin** [comment]
So was the conclusion that the issue of holding back the watermark for in flight data being processed as part of a bundle was specific to FlinkRunner? i.e. other runners do in fact hold back the watermark during the processing of a bundle? I traced the threads and related PRs to the best of my ability, but I wasn't clear on the resolution pertaining to watermark handling during bundles (whereas the thread and PR dealt primarily with timers).  Thanks!

**Comment by egalpin** [comment]
Bump. Are there action items outstanding for this PR or is it blocked on FlinkRunner fix that @lukecwik mentioned?

**Comment by je-ik** [comment]
> Bump. Are there action items outstanding for this PR or is it blocked on FlinkRunner fix that @lukecwik mentioned?

I would not block this. AFAIK this PR preserves the current (although wrong) behavior on flink. I didn't have time to look into the flinkrunner issue so far, but I have it on my list. I think we don't introduce any new bugs by this, we just don't fix it "properly enough" on flink, which is why I would not block. @lukecwik please add your point of view.

**Comment by je-ik** [comment]
There is fix for Flink in https://github.com/apache/beam/pull/17504

**Comment by egalpin** [comment]
force pushing to resolve merge conflict (via rebase)

**Review by je-ik — APPROVED** [review]
LGTM

File: sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java
Comment (lukecwik) [inline sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:2015]:
```suggestion
      /** @deprecated Use {@link #setMaxParallelRequests} instead. */
```

File: sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java
Comment (lukecwik) [inline sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:2178]:
```suggestion
     * ensuring an Elasticsearch cluster is not overwhelmed by parallel requests, but may not work
```

File: sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java
Comment (lukecwik) [inline sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:2181]:
```suggestion
     * runners.
```

File: sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java
Comment (lukecwik) [inline sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:2201]:
```suggestion
     * ensuring an Elasticsearch cluster is not overwhelmed by parallel requests, but may not work
```

File: sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java
Comment (lukecwik) [inline sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:2204]:
```suggestion
     * runners.
```

File: sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java
Comment (lukecwik) [inline sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:2207]:
```suggestion
     * @return the {@link BulkIO} with maximum parallel bulk requests
```

File: sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java
Comment (lukecwik) [inline sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:2512]: nit: it is rare that Java streams perform better and they typically hurt readability

**Comment by egalpin** [comment]
Thanks @je-ik ! Thanks for all your input and guidance along the way with this change (+ related threads and fixes too). Much appreciated!

**Review by lukecwik — APPROVED** [review]
LGTM

I still think there is room for improvement by being able to propagate the original window through and/or using GroupIntoBatches.

**Comment by egalpin** [comment]
thanks @lukecwik ! 

> I still think there is room for improvement by being able to propagate the original window through and/or using GroupIntoBatches.

I agree that can be done by abandoning bundle-based batching[1], but that does necessitate making the ES IO connector non-viable for use on runners without state support.  I'm not clear on how big a deal that is or not ¯\\\_(ツ)_/¯ 

[1] https://github.com/apache/beam/pull/17097

**Pull request merged**
